### PR TITLE
Select all adapters with IPV4 addresses within specified subnet ranges.

### PR DESCRIPTION
**Pull request by drwootton**
Change processing for the --prtemca prte_if_include and --prtemca prte_if_exclude options so that all adapters that are within the specified subnet range are selected or excluded, respectively.

The changes to the split_and_resolve function create a list of unique adapter names selected for processing in the order they are seen, and that list is returned to the caller.

The use case here is: Given a system with three eithernet devices:

eth0: 192.168.1.10
eth1: 192.168.2.10
eth2: 9.10.11.10

All three interfaces can route to other nodes, but we only want to use eth0 and eth1 for TCP traffic for this job.
Currently, I need to specify either the individual devices or non-overlapping subnets: --prtemca prte_if_include 192.168.1.0/24,192.168.2.0/24 because if I specify --prtemca prte_if_include 192.168.0.0/16 only one of the two devices would be selected

With this fix, I can specify -mca btl_tcp_if_include 192.168.1.0/16 and pick up both interfaces.

Related Open MPI Issue: #5818

Signed-off-by: David Wootton <dwootton@us.ibm.com>

### CODE REVIEW

**Comment by jjhursey** [comment]
Ref: https://github.com/open-mpi/ompi/pull/9681

File: src/mca/prteif/base/if_base_components.c [hunk 1]
```diff
@@ -281,35 +293,50 @@ static char **split_and_resolve(char **orig_str, char *name)
...
+                prte_ifindextoname(if_index, if_name, sizeof(if_name));
+                for (n = 0; n < interface_count; n++) {
+                    if (0 == strcmp(if_name, interfaces[n])) {
+                        break;
+                    }
+                }
+                if (n == interface_count) {
+                    prte_output_verbose(20,
+                                        prte_prteif_base_framework.framework_output,
+                                        "oob:tcp: Found match: %s (%s)",
```
[the remark this answers is not in the document]
Comment (drwootton) [reply src/mca/prteif/base/if_base_components.c]: Fixed

**Review by jjhursey — APPROVED** [review]

File: src/mca/prteif/base/if_base_components.c [hunk 2]
```diff
@@ -281,35 +293,50 @@ static char **split_and_resolve(char **orig_str, char *name)
...
-        argv[save++] = strdup(if_name);
-        free(tmp);
     }
 
-    /* The list may have been compressed if there were invalid
-     entries, so ensure we end it with a NULL entry */
-    argv[save] = NULL;
+    /* Mark the end of the interface name array with NULL */
+    interfaces[interface_count] = NULL;
+    free(argv);
```
Comment (rhc54) [inline src/mca/prteif/base/if_base_components.c:338]: Shouldn't this be `prte_argv_free(argv)`?

Comment (drwootton) [reply src/mca/prteif/base/if_base_components.c]: Yes, fixed. Some argv array elements were being freed earlier in the function, so those calls to free are removed.

**Comment by drwootton** [comment]
An OpenMPI Coverity scan found a couple potential problems in source that is almost identical to this fix, so those should be resolved before merging these changes as they likely need to be replicated here.

**Pull request merged**